Feat: post 페이지 기본 구현

ckswnskfk opened the pull request:
## 어떤 변경인지 
- [x] Feat: 기능 변경, 기능 추가
- [ ] Docs: 문서 작업
- [ ] Refactor: 기능 변경 없이 코드 수정
- [ ] Style: 디자인
- [ ] Fix: 버그 수정
- [ ] Test: 테스트 코드 작성

## 설명
pr 올린 이후 rebase 하고 push해서 뭔가 잘못된 pr을 버리고 새 pr을 올립니다..
앞으로 우리 빠르게 머지 해요...!

post 페이지 기본 구현 완성했습니다.
무한 스크롤과 페이지 이동, edit 상태, 미디어 쿼리등은 추후 개발 및 수정하겠습니다.

### 이슈 번호
https://github.com/Codeit-Rolling-11-Letsgo/Rolling/issues/33

### 주요 변경 사항 
1. react router의 loader 사용해 페이지 이동 시 데이터 fetching(공통 fetcher 사용)
2. 용준님께서 만드신 Card 컴포넌트 사용해 PostCard 구현
3. date util 함수 생성

### To Reviewers
셀프 리뷰로 주요 변경 사항과 리뷰 받고 싶은 부분 알려드리겠습니다.

ckswnskfk committed:
Feat: post 페이지 기본 구현

ckswnskfk commented on src/apis/post/postAPI.js:
```diff
@@ -0,0 +1,14 @@
+
+export const getMessages = async ({ recipientId, limit, offset }) => {
+	try {
+		const res = await fetcher.get(
+			`recipients/${recipientId}/messages/?limit=${limit}&offset=${offset}`,
+		);
```
급한대로 이렇게 썼는데 너무 더러워보입니다. 실제 요청도 더럽게 갑니다. 더 깔끔하게 작성하고 보내는 방법이 뭐가 있을까요?

dkile replied:
searchParam 옵션을 활용하면 될 것 같습니다.

ckswnskfk commented on src/components/post/PostCard.jsx:
```diff
@@ -0,0 +1,76 @@
+import styles from '@/components/post/PostCard.module.scss';
+import { formatDate } from '@/utils/dateFunc';
+
+const badgeStyleByRelationship = {
+	지인: 'badgeOrange',
+	동료: 'badgePurple',
+	가족: 'badgeGreen',
+	친구: 'badgeBlue',
+};
+
+function Badge({ relationship }) {
+	return (
+		<span
+			className={clsx(
+				styles.badge,
+				styles[badgeStyleByRelationship[relationship]],
+			)}
+		>
+			{relationship}
+		</span>
+	);
+}
```
이렇게 문자열로 들어오는 인자에 따라 스타일을 다르게 할 경우 더 깔끔하고 안전하게 사용하는 방법은 뭘까요?

mizign replied:
저는 다른 방법으로는 배열 메서드 사용하는거 정도가 생각나는데,, 이것도 충분히 깔끔해보이는 것 같아요

d0ngbb00 replied:
해당 코드만 보면 받는 문자열이 길지 않아서 당장은 그렇게 복잡해보이진 않습니다 !

ckswnskfk commented on src/components/post/PostCard.jsx:
```diff
@@ -0,0 +1,76 @@
+	);
+}
+
+function NewMessage() {
+	return (
+		<Card className={clsx(styles.card, styles.newMessageContainer)}>
+			<Link to={`/post/{id}/message`} className={styles.linkToNewMessage}>
+				<Icon name='plus' className={styles.newMessage} />
+			</Link>
+		</Card>
+	);
+}
+
+PostCard.NewMessage = NewMessage;
```
여긴 이렇게 쓰나 그냥 따로 컴포넌트를 만드나 비슷할 것 같습니다만 좀 따라해봤어요.

dkile replied:
해당 요소는 PostCard의 구성 요소가 아니기 때문에 오히려 헷갈릴 수 있을 것 같습니다. NewMessageCTA 혹은 NewMessageCTACard 정도로 따로 빼는건 어떻게 생각하시나요?

ckswnskfk commented on src/components/post/PostCard.module.scss:
scss가 익숙치 않아 중복을 줄이고 가독성을 높이고 싶은데 쉽지 않네요.

ckswnskfk commented on src/components/post/PostLayout.jsx:
```diff
@@ -0,0 +1,13 @@
+function PostLayout({ children, pageTitle }) {
+	return (
+		<Layout>
+			{pageTitle}
```
여기에 이모지 바가 들어오면 되지 않을까 싶습니다.

mizign commented on src/components/post/PostCard.jsx:
```diff
@@ -0,0 +1,76 @@
+function NewMessage() {
+	return (
+		<Card className={clsx(styles.card, styles.newMessageContainer)}>
+			<Link to={`/post/{id}/message`} className={styles.linkToNewMessage}>
```
혹시 Link의 id부분을 현재 post 페이지의 주인의 id로 처리하려면 NewMessage 컴포넌트도 이를 프롭으로 받아와야하는게 아닐까요? 그리고${id}가 되어야할것 같슴다!

ckswnskfk replied:
넵 맞습니다 이 부분은 url에서 받아온 걸 넣어야겠네요!

dkile commented on src/apis/post/postAPI.js:
```diff
@@ -0,0 +1,14 @@
+			`recipients/${recipientId}/messages/?limit=${limit}&offset=${offset}`,
+		);
+
+		const result = await res.json();
```
이거도 fetcher.get().json()으로 할 수 있어요.

dkile commented on src/apis/post/postAPI.js:
```diff
@@ -0,0 +1,14 @@
+		const result = await res.json();
+		return result;
+	} catch (error) {
+		throw new Error(error);
```
자동으로 HTTPError를 날려주기 때문에 필요없는 구문으로 보입니다.

mizign reviewed (commented):
1. 모바일에서 이렇게 보이는 이슈가 있어서 Layout.module.scss의 이 속성을 비활성화했더니 제대로 보이는 것 같아요 (이건 찬주님 코드는 아니지만 그냥 말씀드려봅니다,,)
![image](https://github.com/Codeit-Rolling-11-Letsgo/Rolling/assets/148179726/468353fb-375f-4a28-8c85-c4e0c58bdf7b)
  ```
.mainContainer {
	height: calc(100% - $size_header);
}
```
2. 뷰포트 360px 이하일때 pc 그리드로 돌아오는 현상이 있습니다
아마 모바일 미디어 쿼리 범위에 min width가 설정되어있어서 그런 것 같아요..!
![image](https://github.com/Codeit-Rolling-11-Letsgo/Rolling/assets/148179726/d12ac5b6-1dee-4924-a751-4003bd042f73)

dkile commented on src/components/post/PostCard.jsx:
```diff
@@ -0,0 +1,76 @@
+	친구: 'badgeBlue',
+};
+
+function Badge({ relationship }) {
```
Badge는 나중에 공통적으로 사용될 수 있는 컴포넌트라서 relationship보다 label 정도가 적절할 것 같네요.

dkile commented on src/main.jsx:
```diff
@@ -18,8 +19,13 @@ const router = createBrowserRouter([
-		path: '/post',
+		path: '/post/:recipientId',
 		element: <PostPage />,
+		loader: ({ params }) =>
```
loader를 잘 쓰셨군요

dkile commented on src/pages/post/PostPage.jsx:
```diff
@@ -1,3 +1,20 @@
+
 export default function PostPage() {
-	return <div>포스트 페이지</div>;
+	const { results } = useLoaderData();
```
results를 그대로 두기에는 어떤 요소를 받아오는지 잘 보이지 않네요.
{results: messageList} 이런식으로 네이밍을 잡아주시면 가독성이 더 좋아질 것 같습니다.

dkile commented on src/utils/dateFunc.js:
```diff
@@ -0,0 +1,8 @@
+export const formatDate = (isoDate) => {
```
Day.js라는 날짜 포매팅 관련 라이브러리도 있는데 추후 복잡한 날짜 포매팅이 필요해진다면 Day.js를 클론해와서 커스텀 해보는 것도 고려해 볼 수 있겠네요

ckswnskfk committed:
Review: 리뷰 사항 반영

api 호출 url params는 URLSearchParams로 깔끔하게 작성
Badge prop명 label로 변경
NewMessageCTA는 따로 컴포넌트로 작성
css 일부 변경

mizign commented:
![image](https://github.com/Codeit-Rolling-11-Letsgo/Rolling/assets/148179726/c0eae064-3929-4164-8527-a65bb1e0b713) 저도 첫번째 리뷰 때 놓친 부분인데, 
플러스 카드 위 패딩값이 피씨, 태블릿, 모바일으로 갈수록 줄어들어야 하는데 고정인것 같아요
이것만 고치면 시안과 완전히 똑같을 것 같아요 💪

ckswnskfk committed:
Review: padding top 조정
